objects.js

```diff
@@ -6164,6 +6211,158 @@ Note.prototype.stop = function () {
     }
 };
 
+Note.prototype.ensureAudioContext = function () {
+    if (AudioContext === undefined) {
+        throw new Error('Web Audio API is not supported\nin this browser');
+    }
+};
+
+// ADSRNote ///////////////////////////////////////////////////////
+
+// I am a single Note with an ADSR envelope that inherits from Note
+ADSRNote.prototype = new Note();
+ADSRNote.prototype.constructor = ADSRNote;
+ADSRNote.uber = Note.prototype;
+ADSRNote.oscMap = {129: "sine", 130: "square", 131: "triangle", 132: "sawtooth"};
+
+// ADSRNote instance creation
+
+function ADSRNote(pitch, instrument, duration) {
+    Note.call(this, pitch, instrument);
+    this.duration = duration;
+    this.noteGain = null;
+}
+
+ADSRNote.prototype.play = function () {
+    this.oscillator = AudioContext.createOscillator();
+    if (!this.oscillator.start) {
+        this.oscillator.start = this.oscillator.noteOn;
+    }
+    if (!this.oscillator.stop) {
+        this.oscillator.stop = this.oscillator.noteOff;
+    }
+
+    // gain for this *specific* note (used to implement envelope)
+    this.noteGain = AudioContext.createGain();
+    this.noteGain.gain.value = 0.0;
+    this.oscillator.start(0); // deprecated, renamed to start()
+
+    // Anchor beginning of ramp at current value.
+    this.noteGain.gain.setTargetAtTime(0.5, AudioContext.currentTime, 0);
+
+    this.oscillator.type = ADSRNote.oscMap[this.instrument];
+    // 0=sin, 1=square, 2=sawtooth, 3=triangle, 4=custom
+
+    this.oscillator.frequency.value =
+        Math.pow(2, (this.pitch - 69) / 12) * 440; // should be 440
+    this.oscillator.connect(this.noteGain);
+    this.noteGain.connect(this.gainNode);
+    this.gainNode.connect(AudioContext.destination);
+};
+
+
+ADSRNote.prototype.stop = function () {
+    Note.call(this);
+    if (this.noteGain) {
+        this.noteGain.gain.setTargetAtTime(0, AudioContext.currentTime, 0);
```

Comment:
I get "SyntaxError: An invalid or illegal string was specified" when I try to play a note (FF 36).